docs/source/common_issues.rst

```diff
@@ -846,3 +846,49 @@ or of potential subclasses.
     # In the context of the subclass Bar, the Self return type promises
     # that the return value will be Bar
     reveal_type(Bar.constructor())  # note: Revealed type is "Bar"
+
+Missing ``yield`` statement in generator functions
+--------------------------------------------------
+
+Python uses the ``yield`` statement to determine whether a function is a
+generator function (https://peps.python.org/pep-0255/#specification-yield).
+
+.. code-block:: python
+
+    from collections.abc import Generator, AsyncGenerator
+
+    async def async_gen() -> AsyncGenerator[int]:
+        yield 1
+
+    def gen() -> Generator[int]:
+        yield 1
+
+    async def async_func() -> int:
+        return 1
+
+    def func() -> int:
+        return 1
+
+    reveal_type(async_gen())  # AsyncGenerator[int, None]
+    reveal_type(gen())        # Generator[int, None, None]
+    reveal_type(async_func()) # Coroutine[Any, Any, int]
+    reveal_type(func())       # int
+
+Functions containing ``yield`` statements will be regarded as Generator functions
+by Python, and their return types are different from other functions.
+Therefore, when using ``mypy`` to check types, we need to declare the return types
+of such functions as Generator (or AsyncGenerator when the function is async).
+
+A common mistake is that we declared the return type of a function as
```

Comment:
And this is wrong: "Missing return statement" is not specific to `Generator`, that's just because the return type is non-None and the body s trivial.

A reasonable function can return a `Generator` or `Iterator` and contain no `yield` (this passes):

```python
from collections.abc import Generator

def fn() -> Generator[int]:
    return (i for i in range(5))
```

On the other hand, missing return statement causes a diagnostic without Generator too (except for stub files):

```python
def fn() -> int:
    ...  # E: Missing return statement  [empty-body]
```

Reply:
That sounds reasonable, I'll re-write these lines.